Add watercliff support to ExperimentalMapGenerator

ashleynewson opened the pull request:
This commit adds watercliffs to generated maps in the RA mod, based on the path partitioning logic added for the D2K map generator.
    
Watercliffs are not available in CnC and unofficial mods, so their use in the ExperimentalMapGenerator configuration is optional.

Depends on #22016, #22017, #22024

ashleynewson committed:
Remove Inner types from various MultiBrush segments

Inner types in MultiBrush segment definitions can usually be derived
automatically from the start and end types. This is especially
important for segments with different starts and ends, where the Inner
type could be either one (e.g. beach to watercliff changers). Inner
types are now only needed in two cases:

- The Inner type doesn't match either the start or end. For example,
  road diagonals.
- The Inner type should only match one of the start or the end. For
  example, cliffs which end into clear tiles.

This commit cleans up the unneeded or unwanted Inner type definitions
in the base mods.

ashleynewson committed:
Adjust point sequences for RA beach segments

The TEMPERAT and SNOW tilesets' MultiBrush segment definitions used
inconsistent conventions for vertical and horizontal beaches. This
meant that:

- Tiling was biased to produce horizontally wider bodies of water.
- Vertical beaches did not line up with watercliffs when connecting.

This change redefines the point sequences to fix these issues.

ashleynewson committed:
Add watercliff support to ExperimentalMapGenerator

This commit adds watercliffs to generated maps in the RA mod, based on
the path partitioning logic added for the D2K map generator.

Watercliffs are not available in CnC and unofficial mods, so their use
in the ExperimentalMapGenerator configuration is optional.

PunkPun reviewed (commented):
'Could not fit tiles for coast' error does seem to happen frequently on dersert water based setups. On some one can barelly find a seed that works.

When trying wetlands I hit an infinite loop inside `PartitionPaths`, it keeps voting and failing. It's the `// Find costs` while(true) loop at line 1366

ashleynewson commented:
> When trying wetlands I hit an infinite loop inside `PartitionPaths`, it keeps voting and failing. It's the `// Find costs` while(true) loop at line 1366

Hmmm... I suspect this isn't an _infinite_ loop, but it looks like the time complexity is much worse than I thought. Trying larger maps does seem to take at least an unreasonable amount of time. I might have to re-think the algorithm here.

Example: Temperate, 256x256, Seed 2020735394, Narrow Wetlands, (others default).

ashleynewson committed:
Optimize Terraformer.PartitionPath

The internal Vote method (called in a tight loop) was inefficient due
to expensive iteration and data structures, which meant that
attempting to generate larger maps could take days.

It is changed to leverage some shared pre-computation to reduce its
runtime complexity and replaces the use of Dictionary with a simple
reused array.

The semantics of the Vote method in isolation are slightly modified,
but this should not impact the quality of generated maps, especially
with current usage.

ashleynewson commented:
PartitionPath optimizations at #22024

ashleynewson commented:
I have modified the constraints and tolerances for partitioning and tiling. This seems to have made coast-related failures much rarer, even on larger maps.

PunkPun commented:
rebase?

PunkPun reviewed (approved):
Seems to work really well

PunkPun commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/22356b8771f7e5273a4dc49e32d5d7a5d08399a9...93a9e648322afab023432d145be0f5acc2873c59
